File: backend/internal/insight/analytics.go
package insight

// Business analytics and forecasting (blueprint D2).
//
// # Every figure here is derived, and none of it is stored
//
// D2 asks for fast movers, dead stock, reorder prediction, a sales forecast,
// profitability and thirteen KPIs. Not one of those is a fact somebody records;
// each is a question about facts that already exist. Materialising them would
// create a second copy of the shop's numbers, and the day it drifted from the
// ledger there would be no way to say which was right.
//
// So this is queries. They are slower than a cached table would be, and that is
// the correct trade for figures an owner makes decisions on.
//
// # Velocity is units a day, over a window somebody chose
//
// The whole module rests on it: fast-moving, dead, reorder date and forecast are
// all velocity read four ways. It is measured over a window the caller names
// rather than all history, because a shirt that sold two hundred last winter and
// none since is not a fast mover, and the average over two years says it is.
//
// # A forecast is arithmetic, and it says so
//
// D2 asks for a "historical-sales-based demand estimate (architecture-ready for
// future ML models)". This is the historical part: velocity times days, with the
// window it was measured over reported alongside. It is not dressed up as a
// prediction, because an owner ordering stock against a number needs to know it
// is last month repeated rather than a model that considered anything else.

import (
	"context"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/shopspring/decimal"

	"github.com/mahedi-emon/rawsyst-pos/backend/internal/platform/db"
)

// Movement is one product and how it has been selling.
type Movement struct {
	VariantID uuid.UUID `json:"variant_id"`
	SKU       string    `json:"sku"`
	Product   string    `json:"product"`
	Category  string    `json:"category,omitempty"`
	Brand     string    `json:"brand,omitempty"`

	SoldQty string `json:"sold_qty"`
	Revenue string `json:"revenue"`
	Profit  string `json:"profit"`
	OnHand  string `json:"on_hand"`

	// Velocity is units sold per day over the window. The number every other
	// figure in this module is read off.
	Velocity string `json:"velocity"`

	// DaysCover is how long the stock on hand lasts at that rate. Null when
	// nothing has sold — a shelf that has not moved does not have a number of
	// days, and reporting a very large one would read as "plenty left".
	DaysCover *int `json:"days_cover,omitempty"`

	// ReorderOn is D2's reorder prediction: the day stock reaches the level
	// the shop set. Empty when there is no reorder level, or nothing is
	// selling, because both mean there is nothing to predict.
	ReorderOn string `json:"reorder_on,omitempty"`

	// DaysSinceSold is what makes something dead stock. −1 when it has never
	// sold at all, which is worse than old and must not sort as recent.
	DaysSinceSold int `json:"days_since_sold"`

	Currency string `json:"currency"`
}

// KPIs are D2's owner dashboard figures.
type KPIs struct {
	From string `json:"from"`
	To   string `json:"to"`

	Revenue     string `json:"revenue"`
	GrossProfit string `json:"gross_profit"`
	GrossMargin string `json:"gross_margin_pct"`

	Orders         int    `json:"orders"`
	AverageOrder   string `json:"average_order_value"`
	UnitsPerOrder  string `json:"units_per_transaction"`
	DiscountRatio  string `json:"discount_ratio_pct"`
	ReturnRate     string `json:"return_rate_pct"`
	InventoryTurn  string `json:"inventory_turnover"`
	RepeatRate     string `json:"repeat_customer_pct"`
	CustomerValue  string `json:"customer_lifetime_value"`
	SalesPerStore  string `json:"sales_per_store"`
	SalesPerPerson string `json:"sales_per_employee"`
	Currency       string `json:"currency"`
}

// Ranked is one category, brand or product by what it actually contributed.
type Ranked struct {
	ID      string `json:"id"`
	Label   string `json:"label"`
	Revenue string `json:"revenue"`
	Cost    string `json:"cost"`
	Profit  string `json:"profit"`
	// Margin is the ratio, which is what ranks a line rather than the absolute
	// profit: a category turning over ten million at two per cent is a
	// different business decision from one turning over a million at thirty.
	Margin   string `json:"margin_pct"`
	Units    string `json:"units"`
	Currency string `json:"currency"`
}

// Movers answers "what is selling and what is not".
//
// One query for both ends of D2's list, because fast-moving and dead stock are
// the same measurement sorted differently, and two queries would be two
// definitions of velocity free to disagree.
func (s *Service) Movers(
	ctx context.Context, scope Scope, days int,
) ([]Movement, error) {
	if days <= 0 || days > 730 {
		days = 90
	}

	out := []Movement{}
	err := s.pool.TxAsTenant(ctx, scope.TenantID, func(tx pgx.Tx) error {
		var currency string
		if e := tx.QueryRow(ctx,
			`SELECT base_currency FROM company WHERE id = $1`,
			scope.CompanyID).Scan(&currency); e != nil {
			return e
		}

		rows, e := tx.Query(ctx, `
			WITH sold AS (
			  SELECT l.variant_id,
			         sum(CASE WHEN i.doc_type = 'credit_note'
			                  THEN -l.qty ELSE l.qty END) AS qty,
			         sum(CASE WHEN i.doc_type = 'credit_note'
			                  THEN -l.net_amount ELSE l.net_amount END) AS revenue,
			         sum(CASE WHEN i.doc_type = 'credit_note'
			                  THEN -l.cogs_amount ELSE l.cogs_amount END) AS cost,
			         max(i.issued_at) FILTER (
			           WHERE i.doc_type <> 'credit_note') AS last_sold
			  FROM sales_invoice_line l
			  JOIN sales_invoice i ON i.id = l.invoice_id
			  WHERE i.company_id = $1
			    AND i.issued_at >= now() - make_interval(days => $2)
			  GROUP BY l.variant_id
			),
			held AS (
			  SELECT m.variant_id, sum(m.delta) AS on_hand
			  FROM stock_movement m
			  WHERE m.company_id = $1
			  GROUP BY m.variant_id
			)
			SELECT v.id, v.sku, p.name,
			       coalesce(c.name, ''), coalesce(b.name, ''),
			       coalesce(s.qty, 0), coalesce(s.revenue, 0),
			       coalesce(s.revenue, 0) - coalesce(s.cost, 0),
			       coalesce(h.on_hand, 0),
			       v.reorder_level,
			       coalesce(extract(day FROM now() - s.last_sold)::int, -1)
			FROM variant v
			JOIN product p ON p.id = v.product_id
			LEFT JOIN category c ON c.id = p.category_id
			LEFT JOIN brand b ON b.id = p.brand_id
			LEFT JOIN sold s ON s.variant_id = v.id
			LEFT JOIN held h ON h.variant_id = v.id
			WHERE v.company_id = $1 AND v.is_active
			ORDER BY coalesce(s.qty, 0) DESC, p.name
			LIMIT 500`, scope.CompanyID, days)
		if e != nil {
			return e
		}
		defer rows.Close()

		window := decimal.NewFromInt(int64(days))
		for rows.Next() {
			var m Movement
			var qty, revenue, profit, onHand decimal.Decimal
			var reorder *decimal.Decimal
			if e := rows.Scan(&m.VariantID, &m.SKU, &m.Product, &m.Category,
				&m.Brand, &qty, &revenue, &profit, &onHand, &reorder,
				&m.DaysSinceSold); e != nil {
				return e
			}

			m.Currency = currency
			m.SoldQty = qty.String()
			m.Revenue = revenue.StringFixed(2)
			m.Profit = profit.StringFixed(2)
			m.OnHand = onHand.String()

			velocity := decimal.Zero
			if qty.IsPositive() {
				velocity = qty.Div(window)
			}
			m.Velocity = velocity.Round(4).String()

			if velocity.IsPositive() && onHand.IsPositive() {
				cover := int(onHand.Div(velocity).IntPart())
				m.DaysCover = &cover

				// D2's reorder prediction. Only when the shop has SET a level:
				// predicting a date against a level nobody chose would be this
				// module inventing the shop's policy.
				if reorder != nil && onHand.GreaterThan(*reorder) {
					until := onHand.Sub(*reorder).Div(velocity)
					m.ReorderOn = time.Now().UTC().
						AddDate(0, 0, int(until.IntPart())).Format("2006-01-02")
				}
			}
			out = append(out, m)
		}
		return rows.Err()
	})
	return out, db.Translate(err, "")
}

// Forecast is D2's demand estimate for one product.
type Forecast struct {
	VariantID uuid.UUID `json:"variant_id"`
	SKU       string    `json:"sku"`
	Product   string    `json:"product"`

	WindowDays   int    `json:"window_days"`
	SoldInWindow string `json:"sold_in_window"`
	Velocity     string `json:"velocity"`

	ForecastDays int    `json:"forecast_days"`
	Expected     string `json:"expected_demand"`
	OnHand       string `json:"on_hand"`
	// Shortfall is what the shop would need to buy to meet it. Zero rather
	// than negative when there is enough: "you are 40 short" and "you have 40
	// spare" are different sentences and this field only says the first.
	Shortfall string `json:"shortfall"`

	// Basis says out loud what this number is. An owner ordering against a
	// forecast has to know it is last month repeated rather than a model that
	// considered a season, a promotion or the weather.
	Basis string `json:"basis"`
}

// Forecasts estimate demand over the coming period.
func (s *Service) Forecasts(
	ctx context.Context, scope Scope, windowDays, forecastDays int,
) ([]Forecast, error) {
	if windowDays <= 0 || windowDays > 730 {
		windowDays = 90
	}
	if forecastDays <= 0 || forecastDays > 365 {
		forecastDays = 30
	}

	movers, err := s.Movers(ctx, scope, windowDays)
	if err != nil {
		return nil, err
	}

	out := make([]Forecast, 0, len(movers))
	for _, m := range movers {
		velocity, _ := decimal.NewFromString(m.Velocity)
		if !velocity.IsPositive() {
			// Nothing sold in the window. A forecast of zero is true and
			// useless, and listing it would bury the products that do move.
			continue
		}
		expected := velocity.Mul(decimal.NewFromInt(int64(forecastDays))).
			Ceil()
		onHand, _ := decimal.NewFromString(m.OnHand)
		shortfall := expected.Sub(onHand)
		if shortfall.IsNegative() {
			shortfall = decimal.Zero
		}

		out = append(out, Forecast{
			VariantID: m.VariantID, SKU: m.SKU, Product: m.Product,
			WindowDays: windowDays, SoldInWindow: m.SoldQty,
			Velocity: m.Velocity, ForecastDays: forecastDays,
			Expected: expected.String(), OnHand: m.OnHand,
			Shortfall: shortfall.String(),
			Basis: "sales over the last " + decimal.NewFromInt(
				int64(windowDays)).String() + " days, repeated",
		})
	}
	return out, nil
}

// Ranking is profitability by category, brand or product.
func (s *Service) Ranking(
	ctx context.Context, scope Scope, by string, from, to time.Time,
) ([]Ranked, error) {
	group := `coalesce(c.id::text, ''), coalesce(c.name, 'Uncategorised')`
	switch by {
	case "brand":
		group = `coalesce(b.id::text, ''), coalesce(b.name, 'No brand')`
	case "product":
		group = `p.id::text, p.name`
	}

	out := []Ranked{}
	err := s.pool.TxAsTenant(ctx, scope.TenantID, func(tx pgx.Tx) error {
		var currency string
		if e := tx.QueryRow(ctx,
			`SELECT base_currency FROM company WHERE id = $1`,
			scope.CompanyID).Scan(&currency); e != nil {
			return e
		}

		// Credit notes are subtracted rather than excluded. A category whose
		// goods mostly come back is not a profitable category, and ranking on
		// gross sales would put it at the top.
		rows, e := tx.Query(ctx, `
			SELECT `+group+`,
			       sum(CASE WHEN i.doc_type = 'credit_note'
			                THEN -l.net_amount ELSE l.net_amount END) AS revenue,
			       sum(CASE WHEN i.doc_type = 'credit_note'
			                THEN -l.cogs_amount ELSE l.cogs_amount END) AS cost,
			       sum(CASE WHEN i.doc_type = 'credit_note'
			                THEN -l.qty ELSE l.qty END) AS units
			FROM sales_invoice_line l
			JOIN sales_invoice i ON i.id = l.invoice_id
			LEFT JOIN variant v ON v.id = l.variant_id
			LEFT JOIN product p ON p.id = v.product_id
			LEFT JOIN category c ON c.id = p.category_id
			LEFT JOIN brand b ON b.id = p.brand_id
			WHERE i.company_id = $1
			  AND i.issued_at >= $2 AND i.issued_at < $3
			GROUP BY 1, 2
			ORDER BY 3 DESC
			LIMIT 100`, scope.CompanyID, from, to)
		if e != nil {
			return e
		}
		defer rows.Close()

		for rows.Next() {
			var r Ranked
			var revenue, cost, units decimal.Decimal
			if e := rows.Scan(&r.ID, &r.Label, &revenue, &cost,
				&units); e != nil {
				return e
			}
			profit := revenue.Sub(cost)
			r.Revenue = revenue.StringFixed(2)
			r.Cost = cost.StringFixed(2)
			r.Profit = profit.StringFixed(2)
			r.Units = units.String()
			r.Currency = currency
			if revenue.IsPositive() {
				r.Margin = profit.Div(revenue).
					Mul(decimal.NewFromInt(100)).StringFixed(1)
			}
			out = append(out, r)
		}
		return rows.Err()
	})
	return out, db.Translate(err, "")
}

File: web-next/src/lib/reports/analytics.ts
// D2's analytics: what sold, what is not moving, and what to order.
//
// # A blank figure is a question this shop cannot answer yet
//
// Four of the thirteen KPIs come back as an empty string on a young business:
// inventory turnover needs a period of purchase history, repeat-customer share
// and lifetime value need customers who have come back. The server sends `""`
// rather than `0`, and the difference is the whole point — a shop told its
// repeat-customer rate is 0% concludes nobody returns, when the truth is that
// nobody has been given time to.
//
// So `stated()` asks whether a figure arrived, never whether it is non-zero,
// and every screen renders an unstated one as a dash with a reason.
//
// # The forecast says what it is
//
// `basis` comes back as "sales over the last 90 days, repeated". It is shown,
// verbatim, next to the number. An owner ordering stock against a forecast has
// to know it is arithmetic on the past and not a prediction — and a forecast
// that hides its method is one people trust more than it deserves.
//
// # Nothing here is stored
//
// Every figure is a question about facts that already exist. Materialising them
// would create a second copy of the shop's numbers, free to drift from the
// ledger they came from.

/** D2's thirteen figures, in one answer so they cannot disagree. */
export interface KPIs {
  from: string;
  to: string;
  currency: string;

  revenue: string;
  gross_profit: string;
  gross_margin_pct: string;
  orders: number;
  average_order_value: string;
  units_per_transaction: string;
  discount_ratio_pct: string;
  return_rate_pct: string;

  /** Empty when the shop has not traded long enough to have one. */
  inventory_turnover: string;
  repeat_customer_pct: string;
  customer_lifetime_value: string;

  sales_per_store: string;
  sales_per_employee: string;
}

/**
 * Whether the server actually stated a figure.
 *
 * Presence, not value. `"0.00"` is a real answer — a month with no returns has
 * a return rate of zero — and treating it as unstated would hide a fact.
 * `""` is the server saying it cannot answer yet.
 */
export function stated(value: string | undefined | null): boolean {
  return typeof value === 'string' && value.trim() !== '';
}

/** One product, measured. Fast and dead stock are the same query, sorted. */
export interface Mover {
  variant_id: string;
  sku: string;
  product: string;
  sold_qty: string;
  revenue: string;
  profit: string;
  on_hand: string;
  velocity: string;
  days_cover?: number;
  reorder_on?: string;
  /** -1 when it has never sold. Not 0, which would mean "sold today". */
  days_since_sold: number;
  currency: string;
}

/**
 * Whether a line is stock that is not moving.
 *
 * `days_since_sold === -1` means it has never sold at all, which is the worst
 * case and the one an owner most wants to see. Anything sitting on the shelf
 * with no sale in the period counts too.
 */
export function isDead(m: Mover): boolean {
  return m.days_since_sold === -1 || Number(m.sold_qty) === 0;
}

/** The two ways the same measurement is read. */
export const MOVER_VIEWS = ['fast', 'dead'] as const;
export type MoverView = (typeof MOVER_VIEWS)[number];

/**
 * The movers, sorted for the question being asked.
 *
 * Fast is by revenue, because "what is selling" means "what is earning" — units
 * sold puts a pile of cheap items above the thing that pays the rent. Dead is
 * by what is tied up in it, for the same reason in reverse.
 */
export function moversFor(rows: Mover[], view: MoverView): Mover[] {
  const alive = rows.filter((m) => !isDead(m));
  const dead = rows.filter(isDead);
  if (view === 'dead') {
    return [...dead].sort((a, b) => Number(b.on_hand) - Number(a.on_hand));
  }
  return [...alive].sort((a, b) => Number(b.revenue) - Number(a.revenue));
}

/** What the next month is expected to need. */
export interface ForecastLine {
  variant_id: string;
  sku: string;
  product: string;
  window_days: number;
  sold_in_window: string;
  velocity: string;
  forecast_days: number;
  expected_demand: string;
  on_hand: string;
  shortfall: string;
  /** The server's own words about what this is. Shown as written. */
  basis: string;
}

/**
 * Whether a line is short, and by enough to act on.
 *
 * The shortfall is the server's arithmetic; this only asks whether it is worth
 * putting in front of somebody. A shortfall of zero is a line that is fine.
 */
export function isShort(line: ForecastLine): boolean {
  const n = Number(line.shortfall);
  return Number.isFinite(n) && n > 0;
}

/** Profit by category, with credit notes subtracted rather than excluded. */
export interface ProfitLine {
  id: string;
  label: string;
  revenue: string;
  cost: string;
  profit: string;
  margin_pct: string;
  units: string;
  currency: string;
}

/** How the workforce is made up, and how much of it is local. */
export interface Workforce {
  total: number;
  saudi: number;
  non_saudi: number;
  saudi_share: string;
  expiring_soon: number;
  expired: number;
  by_department: { department: string; total: number; saudi: number }[];
}

/** A report somebody keeps, and possibly schedules. */
export interface SavedReport {
  id: string;
  name: string;
  kind: string;
  /** A relative phrase, never two dates. */
  period: string;
  store_id?: string;
  warehouse_id?: string;
  account_id?: string;
  cadence?: string;
  day_of_week?: number;
  day_of_month?: number;
  recipients?: string;
  last_run_at?: string;
  last_run_error?: string;
  is_active: boolean;
  /** What the relative period resolves to today, so the screen can show it. */
  from: string;
  to: string;
}

/** The reports that can be kept. The server's own vocabulary. */
export const SAVED_KINDS = [
  'trial_balance',
  'profit_and_loss',
  'balance_sheet',
  'cash_flow',
  'sales',
  'expenses',
  'stock',
  'vat_return',
  'receivables',
  'payables',
  'movers',
  'compliance',
] as const;
export type SavedKind = (typeof SAVED_KINDS)[number];

/**
 * The windows a saved report can cover.
 *
 * Relative, never two dates. "Last month" run in October means September and
 * the same report run in November means October — storing dates would make a
 * saved report a snapshot, and a schedule built on them would email the same
 * figures for ever.
 */
export const SAVED_PERIODS = [
  'today',
  'this_week',
  'this_month',
  'last_month',
  'this_quarter',
  'last_quarter',
  'this_year',
  'last_year',
] as const;
export type SavedPeriod = (typeof SAVED_PERIODS)[number];

/** How often a kept report is sent. Nothing is sent without a schedule. */
export const CADENCES = ['daily', 'weekly', 'monthly'] as const;
export type Cadence = (typeof CADENCES)[number];

/** Why a saved report cannot be kept yet. */
export type SavedProblem =
  | 'no_name'
  | 'no_recipients'
  | 'no_day'
  | 'no_date'
  | 'none';

/**
 * Whether a saved report is ready to keep.
 *
 * The schedule rules are the database's, mirrored: a cadence with nobody to
 * send to "runs every week and reaches nobody", a weekly one needs a day, a
 * monthly one needs a date — and that date stops at 28, because a schedule set
 * for the 31st skips February and a shop that asked for monthly figures
 * quietly gets eleven.
 */
export function savedProblem(draft: {
  name: string;
  cadence: string;
  recipients: string;
  dayOfWeek: string;
  dayOfMonth: string;
}): SavedProblem {
  if (draft.name.trim() === '') return 'no_name';
  if (draft.cadence === '') return 'none';
  if (draft.recipients.trim() === '') return 'no_recipients';
  if (draft.cadence === 'weekly') {
    const d = wholeNumber(draft.dayOfWeek);
    if (d === null || d < 0 || d > 6) return 'no_day';
  }
  if (draft.cadence === 'monthly') {
    const d = wholeNumber(draft.dayOfMonth);
    if (d === null || d < 1 || d > 28) return 'no_date';
  }
  return 'none';
}

/**
 * A whole number, or null when nothing was chosen.
 *
 * `Number('')` is 0, not NaN — so a weekly schedule with no day picked passed
 * every range check as Sunday, and would have been saved to send on a day
 * nobody chose. Caught by a test that expected the empty case to be refused.
 * Zero itself is a real answer here, which is exactly why blank cannot be it.
 */
function wholeNumber(raw: string): number | null {
  const text = raw.trim();
  if (text === '') return null;
  const n = Number(text);
  return Number.isInteger(n) ? n : null;
}

/** The kinds the export route can take away, in its own spelling. */
const EXPORTABLE: Record<string, string> = {
  trial_balance: 'trial-balance',
  profit_and_loss: 'profit-and-loss',
  balance_sheet: 'balance-sheet',
  cash_flow: 'cash-flow',
  vat_return: 'vat-return',
  sales: 'sales',
  expenses: 'expenses',
  stock: 'stock',
};

/**
 * The export path for a saved report's kind, or null.
 *
 * The two vocabularies differ — the saved-report table writes `trial_balance`
 * and the export route takes `trial-balance` — and four savable kinds have no
 * export at all. Returning null rather than guessing a path means the screen
 * shows no button instead of one that 400s.
 */
export function exportKindOf(kind: string): string | null {
  return EXPORTABLE[kind] ?? null;
}
